relay-server/src/services/buffer/envelope_provider/sqlite.rs

```diff
@@ -0,0 +1,628 @@
+    disk_batch_size: usize,
+    max_disk_size: usize,
+}
+
+impl SqliteEnvelopeProvider {
+    /// Creates a new [`SqliteEnvelopeProvider`] instance.
+    pub async fn new(config: &Config) -> Result<Self, SqliteEnvelopeProviderError> {
+        let envelope_store = SqliteEnvelopeStore::prepare(config).await?;
+        Ok(Self {
+            envelopes: HashMap::new(),
+            envelope_store,
+            buffered_envelopes_size: 0,
+            disk_empty: BTreeSet::new(),
+            disk_batch_size: config.spool_envelopes_stack_disk_batch_size(),
+            max_disk_size: config.spool_envelopes_max_disk_size(),
+        })
+    }
+
+    /// Creates a new [`SqliteEnvelopeProvider`] instance given a [`SqliteEnvelopeStore`].
+    pub fn new_with_store(config: &Config, envelope_store: SqliteEnvelopeStore) -> Self {
+        Self {
+            envelopes: HashMap::new(),
+            envelope_store,
+            buffered_envelopes_size: 0,
+            disk_empty: BTreeSet::new(),
+            disk_batch_size: config.spool_envelopes_stack_disk_batch_size(),
+            max_disk_size: config.spool_envelopes_max_disk_size(),
+        }
+    }
+
+    /// Initializes the envelope provider.
+    ///
+    /// Retrieves the project key pairs from the envelope store and creates
+    /// an initialization state.
+    pub async fn initialize(&self) -> InitializationState {
+        match self.envelope_store.project_key_pairs().await {
+            Ok(project_key_pairs) => InitializationState::new(project_key_pairs),
+            Err(error) => {
+                relay_log::error!(
+                    error = &error as &dyn Error,
+                    "failed to initialize the sqlite stack provider"
+                );
+                InitializationState::empty()
+            }
+        }
+    }
+
+    /// Pushes an envelope to the provider for the given project key pair.
+    ///
+    /// If the spool threshold is exceeded, it may trigger spooling to disk.
+    pub async fn push(
+        &mut self,
+        project_key_pair: ProjectKeyPair,
+        envelope: Box<Envelope>,
+    ) -> Result<(), SqliteEnvelopeProviderError> {
+        if self.above_spool_threshold() {
+            self.spool_to_disk().await?;
+        }
+
+        self.envelopes
+            .entry(project_key_pair)
+            .or_default()
+            .push(envelope);
+
+        self.buffered_envelopes_size += 1;
+
+        Ok(())
+    }
+
+    /// Peeks at the next envelope for the given project key pair without removing it.
+    ///
+    /// If no envelope is in the buffer, it will be loaded from disk and a reference will be
+    /// returned.
+    pub async fn peek(
+        &mut self,
+        project_key_pair: ProjectKeyPair,
+    ) -> Result<Option<&Envelope>, SqliteEnvelopeProviderError> {
+        if self.is_empty(project_key_pair) {
+            let envelopes = self.unspool_from_disk(project_key_pair, 1).await?;
+            if envelopes.is_empty() {
```

Comment:
Can `envelopes` even be empty at this point? If not, should we add a `debug_assert!` here?

Reply:
They can theoretically be but if we assume disk is only mutated by a correct process, it will not happen.